Add Multigraph with topological sort and cycle breaking

mikary opened the pull request:
This is part of the fix for #1453. It does not improve the error messages, but with cycle breaking a verbose cycle message may not be necessary
@bricelam @maumar

bricelam commented:
Looks good. Want to take a stab at updating `ModelDiffer` to break cycles? [Here](https://github.com/aspnet/EntityFramework/blob/dev/src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs#L127-L146) is the current implementation.

divega commented on src/EntityFramework.Core/Utilities/Multigraph.cs:
```diff
@@ -0,0 +1,192 @@
+                    {
+                        var canidateVertex = canidateVertices[canidateIndex];
+
+                        // Find verticies in the unsorted portion of the graph that have edges to the canidate
```
typo "canidate"

divega commented:
@mikary This PR just adds some building blocks and doesn't attempt to fix any issues, correct?I would like to see this applied to solving the problem before it gets checked in.

mikary commented:
@bricelam @divega I added basic cycle breaking logic to the ModelDiffer, though it is perhaps overly defensive and I haven't added cycle error messages yet. Is this about what we want the pattern to look like?

bricelam commented on src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs:
```diff
@@ -140,14 +139,27 @@ protected virtual IReadOnlyList<MigrationOperation> Sort([NotNull] IEnumerable<M
+                        }
+                    }
+                    return true;
+                }).Select(k => createTableOperations[k]))
```
This expression got complex. You should probably store the sorted create table operations in a temporary variable before the return statement.

bricelam commented on src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs:
```diff
@@ -140,14 +139,27 @@ protected virtual IReadOnlyList<MigrationOperation> Sort([NotNull] IEnumerable<M
+                {
+                    foreach (var foreignKeyOperation in foreignKeyOperations)
+                    {
+                        if (createTableOperations[key].ForeignKeys.Remove(foreignKeyOperation))
```
Is this ever going to return `false`?

mikary replied:
It shouldn't, the keys are all ones we've already iterated over previously.

bricelam commented:
You should add (at lease) a simple test showing that the model differ can break foreign key cycles.

AndriySvyryd commented on src/EntityFramework.Core/Utilities/Multigraph.cs:
```diff
@@ -0,0 +1,197 @@
+                                {
+                                    sortedQueue.Add(candidateVertex);
+                                    predecessorCounts.Remove(candidateVertex);
+                                    broken = true;
```
You can replace the inner while with a foreach and use break statement

mikary commented:
Updated based on PR feedback and added unit test for ```ModelDiffer``` Also fixed key not found issue #1454 in ```BidirectionalAdjacencyListGraph``` since I was adding comments to that file anyway.

AndriySvyryd commented:
:shipit:

bricelam commented on src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs:
```diff
@@ -140,14 +139,24 @@ protected virtual IReadOnlyList<MigrationOperation> Sort([NotNull] IEnumerable<M
+                    }
+                    return true;
+                })
+                .Select(k => createTableOperations[k]);
```
I'd call `.ToList()` on this. It seems bad to modify the `alterOperations` collection as it lazily evals in the return statement.

mikary replied:
The ```TopologicalSort``` call isn't lazy so ```ToList()``` would only really impact the ```Select(k => createTableOperations[k])``` which is fetching from a dictionary. That being said the lazyness is an implementation detail and not necessarily part of the contract so I could see adding it just to be explicit.
